File: internal/domain/usecase/request/grpc.go
package request

import (
	"context"
	"fmt"
)

const grpcFuncPrefix = "request.usecase"

func (u *usecase) GRPCListServices(ctx context.Context, req GRPCConnectRequest) (*GRPCSchema, error) {
	return u.grpcRequester.ListServices(ctx, req)
}

func (u *usecase) GRPCGenerateExample(ctx context.Context, req GRPCConnectRequest, service, method string) (string, error) {
	schema, err := u.grpcRequester.ListServices(ctx, req)
	if err != nil {
		return "", fmt.Errorf("%s.GRPCGenerateExample: %w", grpcFuncPrefix, err)
	}

	for _, svc := range schema.Services {
		if svc.FullName == service {
			for _, m := range svc.Methods {
				if m.Name == method {
					return m.ExampleJSON, nil
				}
			}
		}
	}

	return "", fmt.Errorf("%s.GRPCGenerateExample: method %s/%s not found", grpcFuncPrefix, service, method)
}

func (u *usecase) GRPCGetProtoDefinition(ctx context.Context, req GRPCConnectRequest, service, method string) (string, error) {
	schema, err := u.grpcRequester.ListServices(ctx, req)
	if err != nil {
		return "", fmt.Errorf("%s.GRPCGetProtoDefinition: %w", grpcFuncPrefix, err)
	}

	for _, svc := range schema.Services {
		if svc.FullName == service {
			for _, m := range svc.Methods {
				if m.Name == method {
					return m.ProtoDefinition, nil
				}
			}
		}
	}

	return "", fmt.Errorf("%s.GRPCGetProtoDefinition: method %s/%s not found", grpcFuncPrefix, service, method)
}

File: frontend/src/types/grpc.ts
export interface GRPCSchema {
  services: GRPCService[]
  source: 'reflection' | 'proto_file' | 'proto_directory'
}

export interface GRPCService {
  fullName: string
  methods: GRPCMethodInfo[]
}

export interface GRPCMethodInfo {
  name: string
  inputType: string
  outputType: string
  isServerStream: boolean
  isClientStream: boolean
  protoDefinition: string
  exampleJson: string
}

export interface GRPCConnectRequest {
  host: string
  useTls: boolean
  protoPath: string
}

export interface GRPCGenerateExampleRequest extends GRPCConnectRequest {
  service: string
  method: string
}
